content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+
+    A copy (d, s) can be emitted safely if:
+    1. Source s won't be overwritten (s is not a dest in remaining copies), AND
+    2. Dest d is not used as a source by another remaining copy
+    """
+    seq = []
+    temp_i = 0
+
+    while copies:
+        dests = {d for (d, s, t) in copies}
+        sources = {s for (d, s, t) in copies}
+
+        # Find a safe copy
+        picked = None
+        for i, (d, s, t) in enumerate(copies):
+            source_safe = (s not in dests) or (s == d)
+            dest_safe = (d not in sources) or (d == s)
+            if source_safe and dest_safe:
+                picked = i
+                break
+
+        if picked is not None:
+            d, s, t = copies.pop(picked)
+            if d != s:
+                seq.append({"op": "id", "dest": d, "type": t, "args": [s]})
+            continue
+
+        # No safe copy - break cycle with temp
+        # Find a value that's both a dest AND a source
+        cycle_var = None
+        for (d, s, t) in copies:
+            if d in sources and d in dests:
+                cycle_var = d
+                cycle_type = t
+                break
+
+        # Save to temp and update copies
+        tmp = f"__phi_tmp{temp_i}"
+        temp_i += 1
+        seq.append({"op": "id", "dest": tmp, "type": cycle_type, "args": [cycle_var]})
+
+        copies = [(d, tmp if s == cycle_var else s, t) for (d, s, t) in copies]
+
+    return seq
+```
+
+#### 2.5.3 The Bug We Fixed
+
+The original algorithm would loop infinitely on certain copy patterns because it didn't correctly identify when to break a cycle. Specifically, it failed when:
+
+1. A destination was used as a source by another copy
+2. Multiple variables formed a cycle
+
+The fix correctly identifies cycle members (variables that are both sources and destinations) and breaks the cycle by saving one to a temporary.
+
+**Example of the fix in action:**
+```
+# Parallel: (x, y, z) := (y, z, x)  # 3-way cycle
+# Step 1: No safe copy (all dests are sources)
+# Step 2: Save x to tmp: tmp = x
+# Step 3: Update: (x, y, z) := (y, z, tmp)
+# Step 4: Now z := tmp is safe, emit it
+# Step 5: Now y := z is safe, emit it
+# Step 6: Now x := y is safe, emit it
+# Result: tmp=x; z=tmp; y=z; x=y
+```
+
+### 2.6 Real Phi Node Representation {#26-real-phi-node-representation}
+
+#### 2.6.1 Why Real Phi Nodes?
+
+The original GVN paper by Cooper and Briggs uses real phi nodes in SSA form. Bril provides an alternative approach using `get`/`set` operations with labels, which is easier to implement but doesn't faithfully represent SSA.
+
+We chose to implement **real phi nodes** for several reasons:
+
+1. **Algorithmic Fidelity**: The GVN algorithm naturally operates on phi nodes as first-class instructions
+2. **Value Numbering of Phis**: We can assign value numbers to phi nodes and detect redundant phis
+3. **Learning Value**: Understanding real SSA conversion is pedagogically important
```

Comment:
This is unsubstantiated. What makes one more "pedagogically important" than the other?